doc/source/ray-core/internals/autoscaler-v2.rst

```diff
@@ -0,0 +1,204 @@
+
+Worker groups can be configured in these ways:
+
+- The `available_node_types <https://docs.ray.io/en/latest/cluster/vms/references/ray-cluster-configuration.html#node-types>`__ field in the Cluster YAML file, if you are using the ``ray up`` cluster launcher.
+- The `workerGroupSpecs <https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/config.html#pod-configuration-headgroupspec-and-workergroupspecs>`__ field in the RayCluster CRD, if you are using KubeRay.
+
+The configuration specifies the logical resources each node has in a worker group, along with the minimum and maximum number of nodes that should exist in each group.
+
+.. note::
+
+   Although the autoscaler fulfills pending resource demands and releases idle nodes, it doesn't perform the actual scheduling of Ray tasks, actors, or placement groups. Scheduling is handled internally by Ray.
+   The autoscaler does its own simulation of scheduling decisions on pending demands periodically to determine which nodes to launch or to stop. See the next sections for details.
+
+
+Periodic Reconciliation
+-----------------------
+
+The entry point of the autoscaler is `monitor.py <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/monitor.py#L332>`__, which starts a GCS client and runs the reconciliation loop.
+
+This process is launched on the head node by the `start_head_processes <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/_private/node.py#L1439>`__ function when using the ``ray up`` cluster launcher.
+When running under KubeRay, it instead runs as a separate autoscaler container in the Head Pod.
+
+.. note::
+
+   In the case of the cluster launcher, if the autoscaler process crashes, then there is no autoscaling.
+   While in the case of KubeRay, Kubernetes restarts the autoscaler container if it crashes.
+
+
+The process periodically `reconciles <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/autoscaler.py#L200-L213>`__ against a snapshot of the following information using the Reconciler:
+
+1. **The latest pending demands** (queried from the `get_cluster_resource_state` GCS RPC): Pending Ray tasks, actors, and placement groups.
+2. **The latest user cluster constraints** (queried from the `get_cluster_resource_state` GCS RPC): The minimum cluster size, if specified via the ``ray.autoscaler.sdk.request_resources`` invocation.
+3. **The latest total and available node resources** (queried from the `get_cluster_resource_state` GCS RPC): The total and currently available resources of each alive Ray node.
+4. **The latest cloud instances** (queried from the cloud provider's implementation): The list of instances managed by the cloud provider implementation.
```

Comment:
cloud provider implementation is the Instance Manager right?

Reply:
Actually, no. Instance Manager is a concrete component that tracks instance lifecycles in memory. It is not an abstract interface that is intended to have multiple implementations. Updated the link to code on this `queried from the cloud provider's implementation` statement, and also added a brief introduction to Instance Manager and Cloud Provider at the top of this section.